packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts

```diff
@@ -0,0 +1,310 @@
+    });
+    return result;
+  };
+
+  /**
+   * Init CloudSDK personalization on server side
+   * @param {IncomingMessage} request incoming nodejs request object
+   * @param {OutgoingMessage} response outgoing nodejs response object
+   * @param {string} [hostname] host for cookies. When not provided, host will be read from host header, and fallback to 'localhost' if that fails
+   */
+  protected async initPersonalizeServer(
+    request: IncomingMessage,
+    response: OutgoingMessage,
+    hostname: string
+  ): Promise<void> {
+    await CloudSDK(request, response, {
+      sitecoreEdgeUrl: this.config.cdpConfig.sitecoreEdgeUrl,
+      sitecoreEdgeContextId: this.config.cdpConfig.sitecoreEdgeContextId,
+      siteName: this.config.sitecoreSiteName,
+      cookieDomain: hostname,
+      enableServerCookie: true,
+    })
+      .addPersonalize({ enablePersonalizeCookie: true })
+      .initialize();
+  }
+
+  protected getVariantIds = async (
+    req: IncomingMessage,
+    language: string,
+    pathname: string
+  ): Promise<string[]> => {
+    // const startTimestamp = Date.now();
+    const timeout = this.config.cdpConfig.timeout;
+
+    // Get personalization info from Experience Edge
+    const personalizeInfo = await this.personalizeService.getPersonalizeInfo(
+      pathname,
+      language,
+      this.config.sitecoreSiteName
+    );
+    if (!personalizeInfo) {
+      // Likely an invalid route / language
+      debug.personalize('skipped (personalize info not found)');
+      return [];
+    }
+
+    if (personalizeInfo.variantIds.length === 0) {
+      debug.personalize('skipped (no personalization configured)');
+      return [];
+    }
+
+    const params = this.getExperienceParams(req);
+    const executions = this.getPersonalizeExecutions(personalizeInfo, language);
+    const identifiedVariantIds: string[] = [];
+    try {
+      await Promise.all(
+        executions.map((execution) =>
+          this.personalize(
+            {
+              friendlyId: execution.friendlyId,
+              variantIds: execution.variantIds,
+              params,
+              language,
+              timeout,
+            },
+            req
+          ).then((personalization) => {
+            const variantId = personalization.variantId;
+            if (variantId) {
+              if (!execution.variantIds.includes(variantId)) {
+                debug.personalize('invalid variant %s', variantId);
+              } else {
+                identifiedVariantIds.push(variantId);
+              }
+            }
+          })
+        )
+      );
+    } catch (e) {
+      debug.personalize('skipped, error %o', e);
+    }
+
+    if (identifiedVariantIds.length === 0) {
+      debug.personalize('skipped (no variant(s) identified)');
+      return [];
+    }
+    return identifiedVariantIds;
+  };
+
+  protected personalizeLayout(layoutData: LayoutServiceData, variantIds: string[]) {
+    if (!layoutData.sitecore?.route) {
```

Comment:
Can we add this check at the very start of the flow? In "personalizeLayoutData" method.
So, we will not be doing anything